optuna/trial/_trial.py

```diff
@@ -485,6 +485,21 @@ def objective(trial):
             )
             raise TypeError(message) from None
 
+        try:
+            if not isinstance(step, int):
+                warnings.warn(
+                    "The reported step is going to cast to int because `step`s type is {}.".format(
+                        type(step).__name__
+                    )
+                )
+                step = int(step)
+
+        except (TypeError, ValueError):
+            message = "The `step` argument is of type '{}' but supposed to be an int.".format(
+                type(step).__name__
+            )
+            raise TypeError(message) from None
```

Comment:
What role does `from None` play in?
Plus, I modified the error message in the PyTorch and Pydantic manner. 
```suggestion
            raise TypeError(f"Expected step as type int, but got {type(step)}.")
```

Reply:
I've just copied this from `value`'s counterpart  https://github.com/optuna/optuna/blob/6aa86fc37811a421d444b6426eae1fc7c544ea96/optuna/trial/_trial.py#L482-L486 for consistency, but I think this `from None` is used not to show the error message made by `int()` according to https://docs.python.org/3/tutorial/errors.html#exception-chaining. 

For example,

### Without `from None`:

```
[I 2024-06-14 14:45:58,473] A new study created in memory with name: no-name-716a04f0-fd8e-4c62-b14b-4127f55dae8f
/Users/nzw0301/Documents/oss/optuna/optuna/trial/_trial.py:490: UserWarning: The reported step is going to cast to int because `step`s type is str.
  warnings.warn(
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
File ~/Documents/oss/optuna/optuna/trial/_trial.py:495, in Trial.report(self, value, step)
    490         warnings.warn(
    491             "The reported step is going to cast to int because `step`s type is {}.".format(
    492                 type(step).__name__
    493             )
    494         )
--> 495         step = int(step)
    497 except (TypeError, ValueError):

ValueError: invalid literal for int() with base 10: '2.2a'

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
Cell In[3], line 4
      1 import optuna
      3 study = optuna.create_study()
----> 4 study.ask().report(2.0, "2.2a")
      5 study.optimize(objective, n_trials=1, n_jobs=1)

File ~/Documents/oss/optuna/optuna/trial/_trial.py:501, in Trial.report(self, value, step)
    497 except (TypeError, ValueError):
    498     message = "The `step` argument is of type '{}' but supposed to be an int.".format(
    499         type(step).__name__
    500     )
--> 501     raise TypeError(message)
    503 if step < 0:
    504     raise ValueError("The `step` argument is {} but cannot be negative.".format(step))

TypeError: The `step` argument is of type 'str' but supposed to be an int.
```

### W/ `from None`

```
[I 2024-06-14 14:46:50,619] A new study created in memory with name: no-name-35a24b5f-4cda-4122-bb9a-1a7c476dc932
/Users/nzw0301/Documents/oss/optuna/optuna/trial/_trial.py:490: UserWarning: The reported step is going to cast to int because `step`s type is str.
  warnings.warn(
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[1], line 4
      1 import optuna
      3 study = optuna.create_study()
----> 4 study.ask().report(2.0, "2.2a")
      5 study.optimize(objective, n_trials=1, n_jobs=1)

File ~/Documents/oss/optuna/optuna/trial/_trial.py:501, in Trial.report(self, value, step)
    497 except (TypeError, ValueError):
    498     message = "The `step` argument is of type '{}' but supposed to be an int.".format(
    499         type(step).__name__
    500     )
--> 501     raise TypeError(message) from None
    503 if step < 0:
    504     raise ValueError("The `step` argument is {} but cannot be negative.".format(step))

TypeError: The `step` argument is of type 'str' but supposed to be an int.
```

the script to generate the error messages is as follows:  

```python
import optuna

study = optuna.create_study()
study.ask().report(2.0, "2.2a")
study.optimize(objective, n_trials=1, n_jobs=1)
```